utils/polus-bfio-util/bfio/bfio.py

```diff
@@ -56,193 +85,746 @@ class BioReader():
+                tile_stride = tile_size.
+                Defaults to None.
+            batch_size (int, optional): Number of tiles to return on each
+                iteration. Defaults to 32.
+            channels (list, optional): A placeholder. Only the first channel
+                is ever loaded. Defaults to [0].
+
+        Yields:
+            numpy.ndarray: A 4-d array where the dimensions are
+                [tile_num,tile_size[0],tile_size[1],channels]
+            list: 
+                
+        Example:
+            from bfio import BioReader
+            import matplotlib.pyplot as plt
+            
+            br = BioReader('/path/to/file')
+            
+            for tiles,ind in br.iterate(tile_size=[256,256],tile_stride=[200,200]):
+                for i in tiles.shape[0]:
+                    print('Displaying tile with X,Y coords: {},{}'.format(ind[i][0],ind[i][1]))
+                    plt.figure()
+                    plt.imshow(tiles[ind,:,:,0].squeeze())
+                    plt.show()
+                
+        """
+        
+        # Enure that the number of tiles does not exceed the width of a supertile
+        if batch_size == None:
+            batch_size = min([32,self.maximum_batch_size(tile_size,tile_stride)])
+        else:
+            assert batch_size <= self.maximum_batch_size(tile_size,tile_stride),\
+                'batch_size must be less than or equal to {}.'.format(self.maximum_batch_size(tile_size,tile_stride))
+        
+        # input error checking
+        assert len(tile_size) == 2, "tile_size must be a list with 2 elements"
+        if tile_stride != None:
+            assert len(tile_stride) == 2, "stride must be a list with 2 elements"
+        else:
+            stride = tile_size
+            
+        # calculate padding if needed
+        if not (set(tile_size) & set(tile_stride)):
+            xyoffset = [(tile_size[0]-tile_stride[0])/2,(tile_size[1]-tile_stride[1])/2]
+            xypad = [(tile_size[0]-tile_stride[0])/2,(tile_size[1]-tile_stride[1])/2]
+            xypad[0] = xyoffset[0] + (tile_stride[0] - np.mod(self.num_y(),tile_stride[0]))/2
+            xypad[1] = xyoffset[1] + (tile_stride[1] - np.mod(self.num_x(),tile_stride[1]))/2
+            xypad = ((int(xyoffset[0]),int(2*xypad[0] - xyoffset[0])),
+                    (int(xyoffset[1]),int(2*xypad[1] - xyoffset[1])))
+        else:
+            xyoffset = [0,0]
+            xypad = ((0,max([tile_size[0] - tile_stride[0],0])),
+                     (0,max([tile_size[1] - tile_stride[1],0])))
+            
+        # determine supertile sizes
+        y_tile_dim = int(np.ceil((self.num_y()-1)/1024))
+        x_tile_dim = 1
+            
+        # Initialize the pixel buffer
+        self._pixel_buffer = np.zeros((y_tile_dim*1024 + tile_size[0],2*x_tile_dim*1024 + tile_size[1]),dtype=self.pixel_type())
+        self._tile_x_offset = -xypad[1][0]
+        self._tile_y_offset = -xypad[0][0]
```

Comment:
We might want to set the tile dimension to be configurable rather than hard coded.

Reply:
Done.